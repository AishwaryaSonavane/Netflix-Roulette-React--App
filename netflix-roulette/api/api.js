export const getMoviesFromApi = (genre='', sortBy='', searchBy='title') => {
   return fetch(`http://localhost:4000/movies?sortBy=${sortBy}&sortOrder=desc&search=${genre}&searchBy=${searchBy}`)
        .then(response => response.json()).then(res => res.data)
}

export const addEditMovieFromApi = (movieData) => {
    return fetch('http://localhost:4000/movies',{
                      method: 'POST',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                      body: JSON.stringify(movieData)
                  })
                  .then(response => response)
}

export const  deleteMovieFromApi = (movieId) => {
      return fetch(`http://localhost:4000/movies/${movieId}`,{
                      method: 'DELETE',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                  })
                  .then(response => response);
}