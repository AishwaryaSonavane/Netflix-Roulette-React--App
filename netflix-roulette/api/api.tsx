import { AddEditMovieFormValues } from "../interface"

export const getMoviesFromApi = (genre='', sortBy='release_date', searchBy='title') => {
   return fetch(`http://localhost:4000/movies?sortBy=${sortBy}&sortOrder=desc&search=${genre}&searchBy=${searchBy}`)
        .then(response => response.json()).then(res => res.data)
}

export const addEditMovieFromApi = (movieData: AddEditMovieFormValues) => {
    return fetch('http://localhost:4000/movies',{
                      method: 'POST',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                      body: JSON.stringify(movieData)
                  })
                  .then(response => response)
}

export const  deleteMovieFromApi = (movieId: number) => {
      return fetch(`http://localhost:4000/movies/${movieId}`,{
                      method: 'DELETE',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                  })
                  .then(response => response);
}