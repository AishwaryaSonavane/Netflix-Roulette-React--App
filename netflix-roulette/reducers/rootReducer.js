import { createSlice } from '@reduxjs/toolkit';
import { HYDRATE } from "next-redux-wrapper";

const initialState = {
    movies: [],
    sortBy: '',
    selectedGenre: '',
    isModalOpen: false,
    editMovieDetails: {},
    selecetedMovie: {},
    searchedMovie: ''
}

export const MovieSlice = createSlice({
  name: 'rootReducer',
  initialState,
  reducers: {
    getMovies: (state,action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.movies = action.payload
    },
    sortMovies: (state,action) => {
        state.sortBy = action.payload
    },
    selectGenre: (state, action) => {
        state.sortBy = action.payload
    },
    openModal: (state) => {
        state.isModalOpen = true
    },
    closeModal: (state) => {
        state.isModalOpen = false
    },
    selecetedMovie: (state, action) => {
        state.selecetedMovie = action.payload
    },
    getSearchedMovie: (state,action) => {
      state.searchedMovie = action.payload
    },
    editMovieAction: (state, action) => {
        state.editMovieDetails = action.payload
    }
  },
  extraReducers: {
    [HYDRATE]: (state, action) => {
      return {
        ...state,
        ...action.payload.rootReducer,
      };
    },
  },
})

// Action creators are generated for each case reducer function
export const { getMovies, getSearchedMovie, openModal, closeModal, selecetedMovie, editMovieAction } = MovieSlice.actions

export const getMoviesData = (state, key) => state.rootReducer.movies;
export const getSelectedSortOption = (state) => state.rootReducer.sortBy;
export const getSelectedGenre = (state) => state.rootReducer.selectedGenre;
export const getModalState = (state) => state.rootReducer.isModalOpen;
export const getEditMovieDetails = (state) => state.rootReducer.editMovieDetails;
export const getSelectedMovie = (state) => state.rootReducer.selecetedMovie;
export const searchedMovie = (state) => state.rootReducer.searchedMovie;

export default MovieSlice.reducer