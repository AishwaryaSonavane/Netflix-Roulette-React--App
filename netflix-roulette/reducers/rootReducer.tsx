import { createSlice } from '@reduxjs/toolkit';
import { HYDRATE } from "next-redux-wrapper";
import { MoviesState, MoviesStateReducer } from '../interface';

const initialState = {
    movies: [],
    sortBy: '',
    selectedGenre: 'All',
    isModalOpen: false,
    editMovieDetails: {},
    selecetedMovie: {},
    searchedMovie: ''
}

export const MovieSlice = createSlice({
  name: 'rootReducer',
  initialState,
  reducers: {
    getMovies: (state,action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.movies = action.payload
    },
    sortMovies: (state,action) => {
        state.sortBy = action.payload
    },
    selectGenre: (state, action) => {
        state.sortBy = action.payload
    },
    openModal: (state) => {
        state.isModalOpen = true
    },
    closeModal: (state) => {
        state.isModalOpen = false
    },
    selecetedMovie: (state, action) => {
        state.selecetedMovie = action.payload
    },
    getSearchedMovie: (state,action) => {
      state.searchedMovie = action.payload
    },
    editMovieAction: (state, action) => {
        state.editMovieDetails = action.payload
    }
  },
  extraReducers: {
    [HYDRATE]: (state, action) => {
      return {
        ...state,
        ...action.payload.rootReducer,
      };
    },
  },
})

// Action creators are generated for each case reducer function
export const { getMovies, getSearchedMovie, openModal, closeModal, selecetedMovie, editMovieAction, selectGenre } = MovieSlice.actions

export const getMoviesData = (state: MoviesStateReducer) => state.rootReducer.movies;
export const getSelectedSortOption = (state: MoviesStateReducer) => state.rootReducer.sortBy;
export const getSelectedGenre = (state: MoviesStateReducer) => state.rootReducer.selectedGenre;
export const getModalState = (state: MoviesStateReducer) => state.rootReducer.isModalOpen;
export const getEditMovieDetails = (state: MoviesStateReducer) => state.rootReducer.editMovieDetails;
export const getSelectedMovie = (state: MoviesStateReducer) => state.rootReducer.selecetedMovie;
export const searchedMovie = (state: MoviesStateReducer) => state.rootReducer.searchedMovie;

export default MovieSlice.reducer