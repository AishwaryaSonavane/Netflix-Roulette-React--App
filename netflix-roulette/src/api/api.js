export function getMoviesFromApi(genre='', sortBy='') {
   return fetch(`http://localhost:4000/movies?sortBy=${sortBy}&sortOrder=desc&search=${genre}&searchBy=genres`)
        .then(response => response.json()).then(res => res.data)
}

export function addEditMovieFromApi(movieData) {
    return fetch('http://localhost:4000/movies',{
                      method: 'POST',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                      body: JSON.stringify(movieData)
                  })
                  .then(response => response)
}

export function deleteMovieFromApi(movieId) {
      return fetch(`http://localhost:4000/movies/${movieId}`,{
                      method: 'DELETE',
                      headers: {
                      'Content-Type': 'application/json',
                      },
                  })
                  .then(response => response);
}