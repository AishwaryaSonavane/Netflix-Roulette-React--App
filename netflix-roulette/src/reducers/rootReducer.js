const initial_state = {
    movies: [],
    sortBy: '',
    selectedGenre: ''
}

const rootReducer = (state = initial_state, action) => {
    switch(action.type) {
        case 'GET_MOVIES': 
            return {
                ...state,
                movies: action.payload
            }
        case 'SORT_MOVIES':  
            return {
                ...state,
                sortBy: action.payload
          }
        case 'SELECT_GENRE':
            return {
                ...state,
                selectedGenre: action.payload
            }  
        default: 
            return state;  
    }
}

export default rootReducer;