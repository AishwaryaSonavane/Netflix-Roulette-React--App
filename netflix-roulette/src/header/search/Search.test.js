import { fireEvent, render as rtlRender, screen, waitFor } from "@testing-library/react"
import { rootReducer } from "../../reducers/rootReducer";
import SearchComponent from "./Search-component";
import { legacy_createStore as createStore} from 'redux';
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { getMoviesFromApi } from "../../api/api";

const store = createStore(rootReducer);

const render = component => rtlRender(
    <Provider store={store}>
        {component}
    </Provider>
)

test('should render search button', () => {
    render(
       <BrowserRouter>
            <SearchComponent/>
       </BrowserRouter>
    );
    const buttonText = screen.getAllByText(/SEARCH/i);
    waitFor(() => expect(buttonText).toBeInTheDocument());
});

test('should match search valu with browser path and action to be called', async () => {
    const spyDispatch = jest.spyOn(store, 'dispatch');

    render(
       <BrowserRouter>
            <SearchComponent/>
       </BrowserRouter>
    );
   
    const input = screen.getByPlaceholderText('What do you want to watch?');

    // Click button
    fireEvent.change(input, {target: {value: 'Coco'}});
    fireEvent.click(screen.getByText('SEARCH'));

    expect(input).toHaveValue('Coco');
    expect(window.location.pathname).toBe('/search/Coco');
    const data = await getMoviesFromApi('Coco','', 'title');
    expect(spyDispatch).toHaveBeenCalled();
});