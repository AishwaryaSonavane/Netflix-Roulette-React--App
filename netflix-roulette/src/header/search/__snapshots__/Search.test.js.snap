// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <input
      class="input"
      placeholder="What do you want to watch?"
    />
    <button
      class="button"
    >
      SEARCH
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <input
      class="input"
      placeholder="What do you want to watch?"
    />
    <button
      class="button"
    >
      SEARCH
    </button>
  </div>
</DocumentFragment>
`;
